src/linux_native.rs

```diff
@@ -0,0 +1,640 @@
+// this
+fn next_hid_usage(cursor: &mut Cursor<&Vec<u8>>, mut usage_page: u16) -> Option<(u16, u16)> {
+    let mut usage = None;
+    let mut usage_pair = None;
+    let initial = cursor.position() == 0;
+
+    while let Some(Ok(key)) = cursor.bytes().next() {
+        // The amount to skip is calculated based off of the start of the
+        // iteration so we need to keep track of that.
+        let position = cursor.position() - 1;
+        let key_cmd = key & 0xfc;
+
+        let (data_len, key_size) = match hid_item_size(key, cursor) {
+            Some(v) => v,
+            None => return None,
+        };
+
+        match key_cmd {
+            // Usage Page 6.2.2.7 (Global)
+            0x4 => {
+                usage_page = hid_report_bytes(cursor, data_len) as u16;
+            }
+            // Usage 6.2.2.8 (Local)
+            0x8 => {
+                usage = Some(hid_report_bytes(cursor, data_len) as u16);
+            }
+            // Collection 6.2.2.4 (Main)
+            0xa0 => {
+                // Usage is a Local Item, unset it
+                if let Some(u) = usage.take() {
+                    usage_pair = Some((usage_page, u))
+                }
+            }
+            // Input 6.2.2.4 (Main)
+		        0x80 |
+            // Output 6.2.2.4 (Main)
+		        0x90 |
+            // Feature 6.2.2.4 (Main)
+		        0xb0 |
+            // End Collection 6.2.2.4 (Main)
+	    0xc0  =>  {
+		// Usage is a Local Item, unset it
+                usage.take();
+            }
+            _ => {}
+        }
+
+        if cursor
+            .seek(SeekFrom::Start(position + (data_len + key_size) as u64))
+            .is_err()
+        {
+            return None;
+        }
+
+        if let Some((usage_page, usage)) = usage_pair {
+            return Some((usage_page, usage));
+        }
+    }
+
+    if let (true, Some(usage)) = (initial, usage) {
+        return Some((usage_page, usage));
+    }
+
+    None
+}
+
+/// Gets the size of the HID item at the given position
+///
+/// Returns data_len and key_size when successful
+fn hid_item_size(key: u8, cursor: &mut Cursor<&Vec<u8>>) -> Option<(usize, usize)> {
+    // Long Item. Next byte contains the length of the data section.
+    if (key & 0xf0) == 0xf0 {
+        if let Some(Ok(len)) = cursor.bytes().next() {
+            return Some((len.into(), 3));
+        }
+
+        // Malformed report
+        return None;
+    }
+
+    // Short Item. Bottom two bits contains the size code
+    match key & 0x03 {
+        v @ 0..=2 => Some((v.into(), 1)),
+        3 => Some((4, 1)),
+        _ => unreachable!(), // & 0x03 means this can't happen
+    }
+}
+
+/// Get the bytes from a HID report descriptor
+///
+/// Must only be called with `num_bytes` 0, 1, 2 or 4.
+fn hid_report_bytes(cursor: &mut Cursor<&Vec<u8>>, num_bytes: usize) -> u32 {
+    let mut bytes: [u8; 4] = [0; 4];
+    if cursor.read_exact(&mut bytes[..num_bytes]).is_err() {
+        return 0;
```

Comment:
This should probably also be returned as an Err/option. And on the call site treat it as a parsing error discarding the whole item?

Reply:
Yes I suppose that's what you should be doing in Rust.